File: backend/models/audit.py
from datetime import datetime
from typing import Any, Literal
import uuid

from pydantic import BaseModel, Field, field_validator

from models.auth import ensure_utc, utc_now


# Actions the browser may record directly; everything else is written server-side only.
ClientAuditAction = Literal[
    "cyclone.selected",
    "satellite.layer_changed",
    "preferences.saved",
    "briefing.exported",
]

AuditScope = Literal["mine", "all"]


class AuditEvent(BaseModel):
    id: str = Field(default_factory=lambda: f"evt_{uuid.uuid4().hex[:16]}")
    actor_user_id: str | None = None
    actor_email: str
    actor_name: str
    actor_role: str
    action: str
    target: str | None = None
    details: dict[str, Any] = Field(default_factory=dict)
    ip: str | None = None
    created_at: datetime = Field(default_factory=utc_now)

    _utc = field_validator("created_at")(ensure_utc)


class AuditEventCreate(BaseModel):
    action: ClientAuditAction
    target: str | None = Field(default=None, max_length=200)
    details: dict[str, Any] = Field(default_factory=dict)


class AuditListResponse(BaseModel):
    events: list[AuditEvent]
    scope: AuditScope
    total: int

File: frontend/src/types/audit.ts
// Mirrors backend/models/audit.py
export type ClientAuditAction = "cyclone.selected" | "satellite.layer_changed" | "preferences.saved" | "briefing.exported";

export type AuditScope = "mine" | "all";

export interface AuditEvent {
  id: string;
  actor_user_id?: string | null;
  actor_email: string;
  actor_name: string;
  actor_role: string;
  action: string;
  target?: string | null;
  details: Record<string, unknown>;
  ip?: string | null;
  created_at: string;
}

export interface AuditEventCreate {
  action: ClientAuditAction;
  target?: string | null;
  details?: Record<string, unknown>;
}

export interface AuditListResponse {
  events: AuditEvent[];
  scope: AuditScope;
  total: number;
}

export interface AuditActionMeta {
  label: string;
  tone: "cyan" | "amber" | "red" | "green" | "muted";
}

// Human labels for every action the backend records (server + client side).
export const AUDIT_ACTIONS: Record<string, AuditActionMeta> = {
  "auth.sign_in": { label: "Signed in", tone: "green" },
  "auth.sign_out": { label: "Signed out", tone: "muted" },
  "auth.sign_in_denied": { label: "Sign-in denied", tone: "red" },
  "session.revoked": { label: "Session revoked", tone: "amber" },
  "session.revoked_others": { label: "Other devices revoked", tone: "amber" },
  "analyst.requested": { label: "Claude analyst requested", tone: "cyan" },
  "admin.domain_added": { label: "Domain approved", tone: "green" },
  "admin.domain_removed": { label: "Domain removed", tone: "red" },
  "admin.role_changed": { label: "Operator role changed", tone: "amber" },
  "cyclone.selected": { label: "Cyclone inspected", tone: "cyan" },
  "satellite.layer_changed": { label: "Satellite layer switched", tone: "cyan" },
  "preferences.saved": { label: "Preferences saved", tone: "muted" },
  "briefing.exported": { label: "Briefing exported", tone: "cyan" },
};

export function describeAction(action: string): AuditActionMeta {
  return AUDIT_ACTIONS[action] ?? { label: action, tone: "muted" };
}
